Support for xcfilelist with project object version >= 50

segiddins opened the pull request:
Closes #8037
Closes https://github.com/CocoaPods/CocoaPods/issues/8181

This is like https://github.com/CocoaPods/CocoaPods/pull/8157, but much less hacky, and with the addition of using a separate file list per each build config.

amorde commented on lib/cocoapods/installer/user_project_integrator/target_integrator.rb:
```diff
@@ -351,20 +391,24 @@ def add_copy_resources_script_phase
+              input_paths_by_config[input_paths_key] = [script_path] + resource_paths
+
+              output_paths_key = [target.copy_resources_script_output_files_path(config), target.copy_resources_script_output_files_relative_path]
+              output_paths_by_config[output_paths_key] = TargetIntegrator.resource_output_paths(resource_paths)
```
could the key be modeled by a `Struct` or something similar? using an Array as a hash key feels a bit confusing to me but I know that is a thing in Ruby land

segiddins replied:
It could be

amorde commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb:
```diff
@@ -83,22 +91,34 @@ def add_copy_resources_script_phase(native_target, test_spec)
-                input_paths = [script_path, *framework_paths.flat_map { |fw| [fw.source_path, fw.dsym_path] }.compact]
-                output_paths = UserProjectIntegrator::TargetIntegrator.framework_output_paths(framework_paths)
+                input_file_list_path = target.embed_frameworks_script_input_files_path_for_test_spec(test_spec)
+                input_file_list_relative_path = "${PODS_ROOT}/#{input_file_list_path.relative_path_from(target.sandbox.root)}"
```
can this use the logic in `AggregateTarget#embed_frameworks_script_input_files_relative_path` somehow? it feels like this logic should be grouped together somewhere. maybe `AggregateTarget` should call out to `PodTargetIntegrator` to get the path

segiddins replied:
not really, because this is for a test spec. this is how we also did the path for the test spec's embed frameworks script

amorde replied:
ah gotcha, totally missed that this is for a test spec

amorde commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb:
```diff
@@ -83,22 +91,34 @@ def add_copy_resources_script_phase(native_target, test_spec)
+                input_paths = input_paths_by_config[input_paths_key] = [script_path]
+                framework_paths.each do |path|
+                  input_paths << path.source_path if path.source_path
+                  input_paths << path.dsym_path if path.dsym_path
```
I'm realizing I probably should have added something like `:all_paths` to the `FrameworkPaths` class for uses like this

segiddins replied:
Can do!

amorde commented on lib/cocoapods/target/pod_target.rb:
```diff
@@ -357,14 +357,32 @@ def test_target_label(test_spec)
 
     # @param  [Specification] test_spec
-    #         The test spec this embed frameworks script path is for.
+    #         The test spec this copy resources script path is for.
```
lol, nice

amorde commented on lib/cocoapods/installer/user_project_integrator/target_integrator.rb:
```diff
@@ -182,10 +222,10 @@ def create_or_update_user_script_phases(script_phases, native_target)
+              phase.shell_path = script_phase[:shell_path]
+              phase.input_paths = script_phase[:input_files]
+              phase.output_paths = script_phase[:output_files]
+              phase.show_env_vars_in_log = script_phase[:show_env_vars_in_log] ? '1' : '0'
```
nice!

amorde commented on lib/cocoapods/installer/user_project_integrator/target_integrator.rb:
```diff
@@ -64,6 +64,48 @@ def initialize(target, installation_options)
+              phase.input_paths = nil
+              phase.output_paths = nil
+              phase.input_file_list_paths = input_paths_by_config.each_key.map(&:last).uniq
+              phase.output_file_list_paths = output_paths_by_config.each_key.map(&:last).uniq
```
another reason for some sort of `Struct` to model this - it's not very clear why `:last` is required, nor what it actually contains

dnkoutso commented on lib/cocoapods/generator/file_list.rb:
```diff
@@ -0,0 +1,39 @@
+        path.open('w') { |file_list| file_list.write(generate) }
+      end
+    end
+  end
```
is formatting correct here?

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb:
```diff
@@ -83,22 +91,34 @@ def add_copy_resources_script_phase(native_target, test_spec)
               end
             end
 
-            UserProjectIntegrator::TargetIntegrator.validate_input_output_path_limit(input_paths, output_paths)
```
no more validation for < 50 ?

segiddins replied:
That's in set_input_output_paths now

dnkoutso commented:
needs rebase, lgtm

dnkoutso commented:
Still LGTM

segiddins commented:
Rebased. In the process, cleaned up some of @dostrander's app spec stuff to avoid duplication (and give myself half as many merge conflicts to resolve)

segiddins commented:
Confirms this works on on project

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:
```diff
@@ -893,32 +893,12 @@ def support_files_group
-          #
-          # @return [Pathname] The absolute path of the Info.plist to use for an app host.
-          #
-          def app_host_info_plist_path_for_test_type(name, test_type)
```
nice

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb:
```diff
@@ -893,32 +893,12 @@ def support_files_group
-          #        The name of the app host.
-          # @return [Pathname] The absolute path of the Info.plist to use for an app host.
-          #
-          def app_host_info_plist_path_for_app_host(name)
```
nice

dnkoutso commented on lib/cocoapods/installer/xcode/pods_project_generator/pod_target_integrator.rb:
```diff
@@ -84,29 +74,33 @@ def add_copy_resources_script_phase(native_target, spec)
-                resource_paths_flattened = resource_paths.flatten.uniq
-                input_paths = [script_path, *resource_paths_flattened]
-                output_paths = UserProjectIntegrator::TargetIntegrator.resource_output_paths(resource_paths_flattened)
+                input_file_list_path = target.embed_frameworks_script_input_files_path_for_spec(spec)
```
I missed this. This is wrong and needs to be `target.copy_resources_script_input_files_path_for_spec(spec)`

Fixing in a PR. This is affects the unreleased 1.7.0

dnkoutso replied:
See https://github.com/CocoaPods/CocoaPods/pull/8222

madnik commented:
What is contributor's opinion about moving similar functionality (That is to move input outputs to xcfilelist in a post-installation hook) as a plugin? Might be useful until we see the release of 1.7.0

segiddins commented:
You're welcome to do that, but our official support will come whenever this patch gets released

Whirlwind commented:
Hi, had this feature been merged into the v1.6.0? I check the Changelog, and it said it is in v1.6.0.beta.2. But I still could not find the source code in the v1.6.0.

dnkoutso commented:
Damn the merging of changelog is wrong fixing.

dnkoutso commented:
Fixed https://github.com/CocoaPods/CocoaPods/pull/8511

revolter commented:
How did you chose to use 50 as the minimum object version? It looks to me that it's available to 46 too.
